k2/csrc/fsa_algo.cc

```diff
@@ -86,30 +101,49 @@ void ConnectCore(const Fsa &fsa, std::vector<int32_t> *state_map) {
 
     const auto &arc = fsa.arcs[current_state.arc_begin];
     auto next_state = arc.dest_state;
-    bool is_visited = visited[next_state];
-    if (!is_visited) {
-      // this is a new discovered state
-      visited[next_state] = true;
-      auto arc_begin = fsa.arc_indexes[next_state];
-      if (next_state != final_state)
-        stack.push({next_state, arc_begin, fsa.arc_indexes[next_state + 1]});
-      else
-        stack.push({next_state, arc_begin, arc_begin});
-
-      if (accessible[current_state.state]) accessible[next_state] = true;
-    } else {
-      // this is a back arc or forward cross arc;
-      // update the co-accessible flag
-      auto next_state = arc.dest_state;
-      if (coaccessible[next_state]) coaccessible[current_state.state] = true;
-      ++current_state.arc_begin;  // go to the next arc
+    auto status = state_status[next_state];
+    switch (status) {
+      case kNotVisited: {
+        // a new discovered node
+        state_status[next_state] = kVisiting;
+        auto arc_begin = fsa.arc_indexes[next_state];
+        if (next_state != final_state)
+          stack.push({next_state, arc_begin, fsa.arc_indexes[next_state + 1]});
+        else
+          stack.push({next_state, arc_begin, arc_begin});
+
+        if (accessible[current_state.state]) accessible[next_state] = true;
+        break;
+      }
+      case kVisiting:
+        // this is a back arc, which means there is a loop in the fsa
+        is_acyclic = false;  // fall-through, no break here
+      case kVisited:
+        // this is a forward cross arc;
+        // update the co-accessible flag
+        if (coaccessible[next_state]) coaccessible[current_state.state] = true;
+        ++current_state.arc_begin;  // go to the next arc
+        break;
+      default:
+        LOG(FATAL) << "Unreachable code is executed!";
+        break;
     }
   }
 
   state_map->reserve(num_states);
+  if (!is_acyclic) {
+    for (auto i = 0; i != num_states; ++i) {
+      if (accessible[i] && coaccessible[i]) state_map->push_back(i);
+    }
+    return;
```

Comment:
I think there should be a return code of some kind here.  E.g. return false if there were cycles found, maybe?

Reply:
It returns true when
- the input fsa is empty
- the output fsa is acyclic
- the outupt is cyclic but contains only self-loops

The output fsa is topsorted for the above three cases.

It returns false otherwise.